### DISCUSSION TITLE
Unable to install/setup wikijs via docker

### DISCUSSION THREAD

**Discussion by uzee001** [requarks/wiki#3989] [Help / Questions]
Hi,
Newbie here but stuck at wikijs installation using docker. I have followed the documentation but I'm not able to set it up. Here's what I have tried:
1. Created an Ubuntu 18 VM and installed docker
2. Installed postgresql and created a db, user with password and granted permissions as mentioned in https://docs.requarks.io/install/requirements.
3. Set and exported the environment variables: DB_HOST, DB_PORT, DB_USER, etc. Did this at the host OS shell level not inside the container.
4. Ran command to start the container as mentioned in https://docs.requarks.io/install/docker:
"docker run -d -p 8080:3000 --name wiki --restart unless-stopped -e "DB_TYPE=postgres" -e "DB_HOST=db" -e "DB_PORT=5432" -e "DB_USER=wikijs" -e "DB_PASS=wikijsrocks" -e "DB_NAME=wiki" requarks/wiki:2"
But could not browse to the url: http://myhost.mydomain:8080, browser says site could not be reached.
5. After some troubleshooting, I found that I need to have postgresql listening on the docker IP (not mentioned in the docs) 
To fix that I modified postgresql.conf to have: listen_addresses = '*'
Once that was done, I now see the following errors in the docker container logs:

```
2021-05-19T16:11:30.753Z [MASTER] info: Connecting to database...
2021-05-19T16:11:30.757Z [MASTER] error: Database Initialization Error: no pg_hba.conf entry for host "172.17.0.2", user "wikijs", database "wiki", SSL off
Loading configuration from /wiki/config.yml... OK
2021-05-19T16:11:31.414Z [MASTER] info: =======================================
2021-05-19T16:11:31.415Z [MASTER] info: = Wiki.js 2.5.201 =====================
2021-05-19T16:11:31.416Z [MASTER] info: =======================================
2021-05-19T16:11:31.416Z [MASTER] info: Initializing...
2021-05-19T16:11:31.974Z [MASTER] info: Using database driver pg for postgres [ OK ]
2021-05-19T16:11:31.977Z [MASTER] info: Connecting to database...
2021-05-19T16:11:31.988Z [MASTER] error: Database Connection Error: 28000 undefined:undefined
2021-05-19T16:11:31.988Z [MASTER] warn: Will retry in 3 seconds... [Attempt 1 of 10]
```
My docker IP is 172.17.0.1, so not sure why the log says 172.17.0.2

Based on the documentation it seemed like its pretty straight forward to install. Unfortunately I don't think thats the case at all, maybe the docs expect new users to be intimately familiar with docker and postgresql. 
If there is something obvious I missed, please let me know. Secondly, if the above is a reasonable attempt at setting up wiki.js using a docker image, then in my humble opinion the docs need to be updated with additional information for brand new users so that they understand the prerequisites properly.
Please advise.
Thanks

**Comment by skledarg** [comment]
The " no pg_hba.conf entry for host" indicates a failed authentication attempt, either because you're connecting with the wrong credentials (wrong user/pass, wrong trust settings for non-local connections...) or your postgres installation is requiring clients to use SSL and you are connecting without it. Btw, 172.17.0.2 seems to be the address of your wiki container, 127.17.0.1 is the gateway for docker network.

From what you wrote it seems you installed the postgres server on the host (as opposed to using the containerized version). In this case, try checking the pg_hba.conf file - try adding this line at the end, if it's not there already (and restart postgres):
`host all all all md5`
 
You might also want to try the dockerized postgres approach as described here: https://docs.requarks.io/install/ubuntu - this might get you up and running easier, without having to touch any postgres config files.

**Comment by uzee001** [comment]
skledarg: thanks very much for your help, sorry it took me a few days to come back to it. Your suggestion to try the dockerized postgres as in https://docs.requarks.io/install/ubuntu worked fine, thanks a lot, much appreciated. Now I have at least a way to get it to work.
But I would also like to learn/understand why the initial approach doesn't work, even after fixing the things you mentioned. Please note that the page https://docs.requarks.io/install/docker and https://docs.requarks.io/install/requirements do not indicate using a dockerized database instance. Instead, it states on the requirements page to have the db setup _"You're expected to have installed one of these database engines already (either locally, on another server or using a cloud service)"_ Hence I had set it up on the local host where docker is running.
As you mentioned I added the line below in pg_hba.conf, and restarted postgres:
`host all all all md5`
I also checked for any ssl configuration in postgres config files and found it in postgresql.conf which I then commented out and restarted postgresql:
```
# grep -i ssl /etc/postgresql/10/main/pg_hba.conf
# hostssl    DATABASE  USER  ADDRESS  METHOD  [OPTIONS]
# hostnossl  DATABASE  USER  ADDRESS  METHOD  [OPTIONS]
# socket, "host" is either a plain or SSL-encrypted TCP/IP socket,
# "hostssl" is an SSL-encrypted TCP/IP socket, and "hostnossl" is a

# grep -i ssl /etc/postgresql/10/main/postgresql.conf
ssl = on
#ssl_ciphers = 'HIGH:MEDIUM:+3DES:!aNULL' # allowed SSL ciphers
#ssl_prefer_server_ciphers = on
#ssl_ecdh_curve = 'prime256v1'
#ssl_dh_params_file = ''
ssl_cert_file = '/etc/ssl/certs/ssl-cert-snakeoil.pem'
ssl_key_file = '/etc/ssl/private/ssl-cert-snakeoil.key'
#ssl_ca_file = ''
#ssl_crl_file = ''
```
After this I got the errors below:
```
# docker run -p 80:3000 --name wiki --restart unless-stopped -e "DB_TYPE=postgres" -e "DB_HOST=172.17.0.2" -e "DB_PORT=5432" -e "DB_USER=wikijs" -e "DB_PASS=wikijsrocks" -e "DB_NAME=wiki" requarks/wiki:2
Loading configuration from /wiki/config.yml... OK
2021-05-23T14:15:00.896Z [MASTER] info: =======================================
2021-05-23T14:15:00.897Z [MASTER] info: = Wiki.js 2.5.201 =====================
2021-05-23T14:15:00.898Z [MASTER] info: =======================================
2021-05-23T14:15:00.898Z [MASTER] info: Initializing...
2021-05-23T14:15:01.474Z [MASTER] info: Using database driver pg for postgres [ OK ]
2021-05-23T14:15:01.478Z [MASTER] info: Connecting to database...
2021-05-23T14:15:01.484Z [MASTER] error: Database Connection Error: ECONNREFUSED 172.17.0.2:5432
2021-05-23T14:15:01.485Z [MASTER] warn: Will retry in 3 seconds... [Attempt 1 of 10]
```
I searched on the error and thought its the config.yml file causing it, I found a config.yml file at this path in my VM:
`/var/lib/docker/overlay2/4719ea426e48e1159830ecb1d411db7c97d77c368ec894b07c74b35f428a27b6/diff/wiki/config.yml`
This had references to ssl and lets_encrypt which made me think it should be removed and a new config.yml will be created when I run the "docker run" command again, unfortunately thats not the case and now its giving this error:
```
ENOENT: no such file or directory, open '/wiki/config.yml'
>>> Unable to read configuration file! Did you create the config.yml file?
Loading configuration from /wiki/config.yml... FAILED
```

My mistake that I didn't save the file, but I know that I never created it and based on the path I thought it was created by the docker run command as I had not done anything else with docker besides just installing the packages and then following the wiki.js installation page.

Sorry for the long post, but would really appreciate any help/advice so that I can understand why this is not working.
Thanks.
